### PR TITLE
add an error injection executor

### PR DESCRIPTION
**Pull request by phlogistonjohn**
### What does this PR achieve? Why do we need it?

This series adds two executors "stack" and "injectexec". The stack executor simply supports stacking executors together (this may be split out into a seperate PR). The inject executor supports error injection by specifying "inject/<otherexecutor>" where currently ssh and mock are supported.

I plan on using this framework to test the functionality of pending operation cleanup. In other words, the ability for heketi to automatically clean up a pending operation such as a volume delete after a crash. In order to properly test that we need a way to do a "controlled crash" this (preview) PR is the beginning of that.

### Notes for the reviewer

This series aims to gain a basic infrastructure to do error injection, mostly around commands, so hat the functional test framework has the ability to trigger various error condition and test the clean code that currently exists and more importantly more complex cleanup code that will be developed. However, this series only aims at getting the executor itself, and some validation that it works as designed.

To get a quick taste of how I imagine this will be used, picture a test that will create a heketi server with a custom configuration file. The json will specify an executor such as "inject/ssh", which stacks the error injection layer on top of the normal ssh executor.
Then a new section of json will specify what sorts of things to inject:
```
       "injectexec": {
          "command_injection": {
               "commands": [
                   {"Cmd": "lvcreate.*", "Reaction": {"Panic": "FOO!"}}
               ]
           }
       }
```
In this case we're requesting the system to panic *before* any lvcreate command is called on the underlying system.

This series has been rebased on top of the changes in PR #1317 in order to re-use common code for test setup and teardown and server management. Please review and merge that series prior to this one.

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Patches rebased and cleaned up a bit. This still isn't in a done state, however it's much more readable now.

**Comment by phlogistonjohn** [comment]
I feel that these patches are now in a reviewable state.

**Comment by phlogistonjohn** [comment]
I'm adding back the do not merge tag in order to signal that this needs to be rebased upon the cleanups to the executor apis that are now ongoing. In the long term this is a good thing because the error injection framework will fit better with this api changes.

**Comment by phlogistonjohn** [comment]
I have rebased and removed the "do not merge" label once again.
Please review with priority as this will be needed to test the changes coming in for issue #1401

**Comment by phlogistonjohn** [comment]
retest this please

**Comment by phlogistonjohn** [comment]
Updated vagrant setup stuff to enable the use of gluster block. When this pr was started the TestSmokeTest vagrant dir, that this is based off of did not have block support. Now that it does, this change harmonizes the setups.

**Comment by phlogistonjohn** [comment]
@obnoxxx @raghavendra-talur @Madhu-1 @nixpanic @ansiwen PTAL
Getting this one reviewed & merged will significantly accelerate the pending ops cleanup work.

File: executors/injectexec/hooks_test.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+//
+
+package injectexec
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/heketi/tests"
+
```
Comment (Madhu-1) [inline executors/injectexec/hooks_test.go]: this new line can be removed

File: executors/injectexec/injectexec.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	ie.Pre = newMockBase()
+	ie.realExecutor = e
+
+	// set up the exec stack to include our dummy "pre"-executor
+	// and the real executor
+	ie.SetExec([]executors.Executor{ie.Pre, ie.realExecutor})
+
+	// set up wrapping of real executors that are command based
+	// to load the hooks for individual commands
+	switch x := e.(type) {
```
Comment (Madhu-1) [inline executors/injectexec/injectexec.go]: can we use fallthough as below

```
switch x := e.(type) {
	case *sshexec.SshExecutor:
		fallthrough
	case *kubeexec.KubeExecutor:
		fallthrough
	case *cmdexec.CmdExecutor:
		logger.Info("injecting executor with transport")
		ie.realTransport = x.RemoteExecutor
		x.RemoteExecutor = ie.Wrap(x.RemoteExecutor)
	default:
		logger.Info("not wrapping executor without known transport")
	}
	logger.Info("created inject executor (from %T)", ie.realExecutor)
	return ie

```

File: executors/injectexec/injectexec.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	case *kubeexec.KubeExecutor:
+		logger.Info("injecting executor with transport")
+		ie.realTransport = x.RemoteExecutor
+		x.RemoteExecutor = ie.Wrap(x.RemoteExecutor)
+	case *cmdexec.CmdExecutor:
+		logger.Info("injecting executor with transport")
+		ie.realTransport = x.RemoteExecutor
+		x.RemoteExecutor = ie.Wrap(x.RemoteExecutor)
+	default:
+		logger.Info("not wrapping executor without known transport")
```
Comment (Madhu-1) [inline executors/injectexec/injectexec.go]: this should be `Info` or `warning` or something else?

File: executors/injectexec/injectexec_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+package injectexec
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/heketi/heketi/executors/cmdexec"
+	"github.com/heketi/heketi/executors/mockexec"
+
```
Comment (Madhu-1) [inline executors/injectexec/injectexec_test.go]: can we remove this extra line?

File: executors/injectexec/injectexec.go (hunk 3 above)
Comment (phlogistonjohn) [reply executors/injectexec/injectexec.go]: A warning seems reasonable.

File: executors/injectexec/hooks_test.go (hunk 1 above)
Comment (phlogistonjohn) [reply executors/injectexec/hooks_test.go]: I am trying to group imports by origin, first by std library, then by external dependency, then by internal (in-repo) package. The "heketi/tests" package is not part of "heketi/heketi" and that's why its by itself in a single line group in some of these.

File: executors/injectexec/injectexec_test.go (hunk 4 above)
Comment (phlogistonjohn) [reply executors/injectexec/injectexec_test.go]: I would rather just fix the ordering mistake here.

File: executors/injectexec/transport_test.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	w1.handleBefore = func(c string) rex.Result {
+		return rex.Result{
+			Completed: true,
+			Err:       fmt.Errorf("HipHooray"),
+		}
+	}
+	r, err = w2.ExecCommands("foo", []string{"ls 1", "ls 2"}, 10)
+	tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	tests.Assert(t, !r.Ok(), "epxected r.Ok() is false")
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply executors/injectexec/transport_test.go]: Right. :-)

File: executors/injectexec/injectexec.go (hunk 2 above)
Comment (phlogistonjohn) [reply executors/injectexec/injectexec.go]: Nope. It's not permitted in a type switch.

**Pull request merged**